src/server/db/actions/GameAction.ts

```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-    const result = await GameModel.findByIdAndUpdate(allData.id, allData.data, {
-      new: false,
+  let total_tags: ObjectId[] = [];
+  //Find the correspoding tag ids for the accessibility TAGS.
+  if (query.accessibility) {
+    const accessibility_tags = await TagModel.find({
+      name: { $in: query.accessibility },
+      type: "accessibility",
     });
-    if (!result) {
-      throw new GenericUserErrorException("Game with given ID does not exist.");
+    //Ensure that we found all tags and no extras
+    if (query.accessibility.length !== accessibility_tags.length) {
+      throw new GenericUserErrorException(
+        "One or more of the accessibility tags do not exist.",
+      );
     }
-    return result;
-  } catch (e) {
-    throw e;
+    //Filter used to ensure object is not null, then map converts it to an array of object ids.
+    total_tags = total_tags.concat(
+      accessibility_tags.filter((tag) => tag).map((tag) => tag.id),
+    );
   }
-}
-
-export async function getAllGames() {
-  await connectMongoDB();
-  try {
-    const games = await GameModel.find();
-    if (games == null) {
-      return [];
+  if (query.tags) {
+    const normal_tags = await TagModel.find({ name: { $in: query.tags } });
+    if (query.tags.length !== normal_tags.length) {
+      throw new GenericUserErrorException(
+        "One or more of the custom tags do not exist.",
+      );
     }
-    return games;
-  } catch (e) {
-    throw e;
+    total_tags = total_tags.concat(
+      normal_tags.filter((tag) => tag).map((tag) => tag.id),
+    );
   }
+  if (total_tags.length !== 0) {
+    //Add to filters
+    filters.tags = { $all: total_tags };
+  }
+  //Currently ignoring filtering by build, as that hasn't yet been implemented.
+
+  //Match by name with regex
+  if (query.name && query.name !== "") {
+    const reg_string = new RegExp(query.name, "i");
+    filters.name = { $regex: reg_string };
+  }
+  //Match by gameContent
+  if (query.gameContent && "answerKey" in query.gameContent) {
+    filters.answerKey = { $exists: true };
+  }
+  if (query.gameContent && "parentingGuide" in query.gameContent) {
+    filters.parentingGuide = { $exists: true };
+  }
+  if (query.gameContent && "lessonPlan" in query.gameContent) {
+    filters.lessonPlan = { $exists: true };
+  }
+  if (query.gameContent && "videoTrailer" in query.gameContent) {
+    filters.videoTrailer = { $exists: true };
+  }
+  let games;
+  if (query.page) {
+    games = await GameModel.find(filters)
+      .skip((query.page - 1) * RESULTS_PER_PAGE)
+      .limit(RESULTS_PER_PAGE);
+  } else {
+    games = await GameModel.find(filters);
+  }
+  return games;
 }
 
 export async function getGameById(id: string) {
   await connectMongoDB();
-  try {
-    const game = await GameModel.findById(id)
-      .populate("themes")
-      .populate("tags");
-    return game;
-  } catch (e) {
-    throw e;
-  }
+  const game = await GameModel.findById(id).populate("themes").populate("tags");
```

Comment:
After changing the schema as discussed earlier, pass in a type into `populate<...>()`